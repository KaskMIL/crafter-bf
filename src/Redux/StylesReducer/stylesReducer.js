import { v4 as uuidv4 } from 'uuid';

// URLs
const stylesURL = 'https://rustybeer.herokuapp.com/styles';
// const hopsURL = 'https://rustybeer.herokuapp.com/hops';
// const yeastsURL = 'https://rustybeer.herokuapp.com/yeasts';

// Actions
const GETSTYLES = 'crafter-bf/redux/GET_STYLES';
// const HOPS = 'crafter-bf/redux/GET_HOPS';
// const YEASTS = 'crafter-bf/redux/GET_YEASTS';

// Action creators

export const getStylesList = () => async (dispatch) => {
  const response = await fetch(stylesURL, {
    method: 'GET',
  });

  const styleData = await response.json();
  const dataArr = styleData.map((beer) => {
    const newBeer = {
      name: beer.name,
      id: uuidv4(),
      abv_min: beer.abv_min,
      abv_max: beer.abv_max,
      ibu_min: beer.ibu_min,
      ibu_max: beer.ibu_max,
      description: beer.description,
    };
    return newBeer;
  });

  dispatch({
    type: GETSTYLES,
    payload: dataArr,
  });
};

// Reducer

const stylesReducer = (state = [], action) => {
  switch (action.type) {
    case GETSTYLES:
      return action.payload;
    default:
      return state;
  }
}

export default stylesReducer;