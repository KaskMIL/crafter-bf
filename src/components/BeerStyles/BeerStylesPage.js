import React, { useEffect, useState } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import Slider from '@mui/material/Slider';
import { getStylesList, showInfo } from '../../Redux/StylesReducer/stylesReducer';
import BeerStylesHeader from '../ElementHeader/BeerStylesHeader';
import ButtonsContainer from '../ButtonsContainer/ButtonsContanier';
import BeerStyleElement from '../ElementComponent/BeerStyleElement';
import styles from './BeerStylesPage.module.scss';

function BeerStylesPage() {
  // Set state to manipulate ABV data
  const [abvValue, setAbvValue] = useState({
    value: [1, 14],
    active: false,
  });

  // Set state to manipulate IBU data
  const [ibuValue, setIbuValue] = useState({
    value: [0, 120],
    active: false,
  });

  // Set dispatch and useSelector
  const beerList = useSelector((state) => state.styles);
  const dispatch = useDispatch();

  // useEffect to get data from the API
  useEffect(() => {
    if (!beerList.length) {
      dispatch(getStylesList());
    }
  }, [dispatch, beerList]);

  // Function to change the value of ABV state
  const handleAbvChange = (newValue) => {
    setAbvValue({ ...abvValue, value: newValue });
  };

  // Function to invert ABV active value and set value to default
  const handleAbvButton = () => {
    if (!abvValue.active && !ibuValue.active) {
      setAbvValue({ ...abvValue, active: !abvValue.active });
    } else if (!abvValue.active && ibuValue.active) {
      setAbvValue({ ...abvValue, active: !abvValue.active });
      setIbuValue({ ...ibuValue, active: !ibuValue.active });
    } else {
      setAbvValue({
        ...abvValue,
        value: [1, 10],
        active: false,
      });
    }
  };

  // Function to change value of IBU state
  const handleIbuChange = (newValue) => {
    setIbuValue({ ...ibuValue, value: newValue });
  };

  // Function to invert IBU active value  and set value to default
  const handleIbuButton = () => {
    if (!ibuValue.active && !abvValue.active) {
      setIbuValue({ ...ibuValue, active: !ibuValue.active });
    } else if (!ibuValue.active && abvValue.active) {
      setIbuValue({ ...ibuValue, active: !ibuValue.active });
      setAbvValue({ ...abvValue, active: !abvValue.active });
    } else {
      setIbuValue({
        ...ibuValue,
        value: [0, 120],
        active: !ibuValue.active,
      });
    }
  };

  const handleReducer = (id) => {
    dispatch(showInfo(id));
  };

  const getText = (value, string) => `${value} ${string}`;

  return (
    <main className={styles.container}>
      <BeerStylesHeader title="Beer Styles" />
      <ButtonsContainer
        handleAbvButton={handleAbvButton}
        handleIbuButton={handleIbuButton}
        firstBtnTitle="ABV Filter"
        secondBtnTitle="IBU Filter"
      />
      <div className={styles.sliderContainer}>
        {
          abvValue.active ? (
            <h2>ABV Filter</h2>
          ) : null
        }
        {
          ibuValue.active ? (
            <h2>IBU Filter</h2>
          ) : null
        }
        {
          // ABV //
          abvValue.active ? (
            <Slider
              onChange={(e) => handleAbvChange(e.target.value)}
              getAriaLabel={() => 'ABV'}
              valueLabelDisplay="auto"
              getAriaValueText={() => getText(abvValue.value, 'ABV')}
              value={abvValue.value}
              step={1}
              marks
              min={1}
              max={14}
              sx={{
                width: 300,
                color: 'white',
              }}
            />
          ) : null
        }
        {
          // IBU //
          ibuValue.active ? (
            <Slider
              onChange={(e) => handleIbuChange(e.target.value)}
              getAriaLabel={() => 'IBU'}
              valueLabelDisplay="auto"
              getAriaValueText={() => getText(ibuValue.value, 'IBU')}
              value={ibuValue.value}
              step={1}
              marks
              min={0}
              max={120}
              sx={{
                width: 300,
                color: 'white',
              }}
            />
          ) : null
        }
      </div>
      <ul>
        {beerList.map((beer) => {
          // ABV Filter
          if (abvValue.active) {
            if (
              abvValue.value[0] <= beer.abv_max
              && abvValue.value[1] >= beer.abv_max
            ) {
              return (
                <BeerStyleElement
                  key={beer.id}
                  name={beer.name}
                  type="beer"
                  id={beer.id}
                  firstTitle="ABV:   "
                  firstMax={beer.abv_max}
                  firstMin={beer.abv_min}
                  secondTitle="IBU:   "
                  secondMax={beer.ibu_max}
                  secondMin={beer.ibu_min}
                  description={beer.description}
                  show={beer.show}
                  reducer={handleReducer}
                />
              );
            }
            return null;
          }
          // IBU Filter
          if (ibuValue.active) {
            if (
              beer.ibu_max >= ibuValue.value[0]
              && beer.ibu_max <= ibuValue.value[1]
            ) {
              return (
                <BeerStyleElement
                  key={beer.id}
                  id={beer.id}
                  type="beer"
                  name={beer.name}
                  firstTitle="ABV:   "
                  firstMax={beer.abv_max}
                  firstMin={beer.abv_min}
                  secondTitle="IBU:   "
                  secondMax={beer.ibu_max}
                  secondMin={beer.ibu_min}
                  description={beer.description}
                  show={beer.show}
                  reducer={handleReducer}
                />
              );
            }
            return null;
          }
          return (
            <BeerStyleElement
              key={beer.id}
              name={beer.name}
              type="beer"
              id={beer.id}
              firstTitle="ABV   "
              firstMax={beer.abv_max}
              firstMin={beer.abv_min}
              secondTitle="IBU   "
              secondMax={beer.ibu_max}
              secondMin={beer.ibu_min}
              description={beer.description}
              show={beer.show}
              reducer={handleReducer}
            />
          );
        })}
      </ul>
    </main>
  );
}

export default BeerStylesPage;
