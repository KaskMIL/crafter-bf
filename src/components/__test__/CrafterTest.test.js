import {
  render, screen, waitFor,
} from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import * as actions from '../../Redux/StylesReducer/stylesReducer';
import store from '../../Redux/configStore';
import BeerStylesPage from '../BeerStyles/BeerStylesPage';

describe('Testing the reducers', () => {
  test('Unknown action.type returns the current state', () => {
    const initial = [
      {
        name: 'Lite American Lager',
        id: 'b5fca833-cc25-4c9c-8ce6-dd93fb3c6a98',
        abv_min: 2.8,
        abv_max: 4.2,
        ibu_min: 8,
        ibu_max: 12,
        description:
          'Highly carbonated, very light-bodied, nearly flavorless lager designed to be consumed very cold. Very refreshing and thirst quenching.',
        show: false,
      },
    ];
    const unknownAction = () => ({ type: 'unknown', payload: 'something' });
    expect(actions.default(initial, unknownAction)).toEqual(initial);
  });
  test('Change status active false to true and viceversa', () => {
    const initial = [
      {
        name: 'Lite American Lager',
        id: 1,
        abv_min: 2.8,
        abv_max: 4.2,
        ibu_min: 8,
        ibu_max: 12,
        description:
          'Highly carbonated, very light-bodied, nearly flavorless lager designed to be consumed very cold. Very refreshing and thirst quenching.',
        show: false,
      },
      {
        name: 'Lite American Lager',
        id: 2,
        abv_min: 2.8,
        abv_max: 4.2,
        ibu_min: 8,
        ibu_max: 12,
        description:
          'Highly carbonated, very light-bodied, nearly flavorless lager designed to be consumed very cold. Very refreshing and thirst quenching.',
        show: false,
      },
      {
        name: 3,
        id: 'b5fca833-cc25-4c9c-8ce6-dd93fb3c6a98',
        abv_min: 2.8,
        abv_max: 4.2,
        ibu_min: 8,
        ibu_max: 12,
        description:
          'Highly carbonated, very light-bodied, nearly flavorless lager designed to be consumed very cold. Very refreshing and thirst quenching.',
        show: false,
      },
    ];
    const final = [
      {
        name: 'Lite American Lager',
        id: 1,
        abv_min: 2.8,
        abv_max: 4.2,
        ibu_min: 8,
        ibu_max: 12,
        description:
          'Highly carbonated, very light-bodied, nearly flavorless lager designed to be consumed very cold. Very refreshing and thirst quenching.',
        show: false,
      },
      {
        name: 'Lite American Lager',
        id: 2,
        abv_min: 2.8,
        abv_max: 4.2,
        ibu_min: 8,
        ibu_max: 12,
        description:
          'Highly carbonated, very light-bodied, nearly flavorless lager designed to be consumed very cold. Very refreshing and thirst quenching.',
        show: true,
      },
      {
        name: 3,
        id: 'b5fca833-cc25-4c9c-8ce6-dd93fb3c6a98',
        abv_min: 2.8,
        abv_max: 4.2,
        ibu_min: 8,
        ibu_max: 12,
        description:
          'Highly carbonated, very light-bodied, nearly flavorless lager designed to be consumed very cold. Very refreshing and thirst quenching.',
        show: false,
      },
    ];
    expect(actions.default(initial, actions.showInfo(2))).toEqual(final);
  });
});

describe('Body of Styles', () => {
  test('Check Lite American Lager appears in DOM', async () => {
    render(
      <MemoryRouter>
        <Provider store={store}>
          <BeerStylesPage />
        </Provider>
      </MemoryRouter>,
    );
    await waitFor(() => {
      const text = screen.findByText(
        /Highly carbonated, very light-bodied, nearly flavorless lager designed to be consumed very cold. Very refreshing and thirst quenching./,
      );
      expect(text).not.toBe(null);
    });
  });
  test('Matches snapshot', () => {
    const tree = renderer.create(
      <MemoryRouter>
        <Provider store={store}>
          <BeerStylesPage />
        </Provider>
      </MemoryRouter>,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
