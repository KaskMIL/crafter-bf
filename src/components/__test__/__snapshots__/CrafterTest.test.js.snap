// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Body of Styles Matches snapshot 1`] = `
<main
  className="container"
>
  <div
    className="headerContainer"
  >
    <a
      href="/"
      onClick={[Function]}
    >
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        strokeWidth="0"
        style={
          Object {
            "color": undefined,
          }
        }
        viewBox="0 0 512 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M217.9 256L345 129c9.4-9.4 9.4-24.6 0-33.9-9.4-9.4-24.6-9.3-34 0L167 239c-9.1 9.1-9.3 23.7-.7 33.1L310.9 417c4.7 4.7 10.9 7 17 7s12.3-2.3 17-7c9.4-9.4 9.4-24.6 0-33.9L217.9 256z"
        />
      </svg>
    </a>
    <h1>
      Beer Styles
    </h1>
  </div>
  <div
    className="btnContainer"
  >
    <button
      onClick={[Function]}
      type="submit"
    >
      ABV Filter
    </button>
    <button
      onClick={[Function]}
      type="submit"
    >
      IBU Filter
    </button>
  </div>
  <div
    className="sliderContainer"
  />
  <ul />
</main>
`;
